compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/inference/PostponedArgumentsAnalyzer.kt

```diff
@@ -164,21 +164,30 @@ class PostponedArgumentsAnalyzer(
 
         val checkerSink: CheckerSink = CheckerSinkImpl(candidate)
 
+        val lastExpression = lambda.atom.body?.statements?.lastOrNull() as? FirExpression
         var hasExpressionInReturnArguments = false
+        // Don't constraint for return expressions of lambda if it has Unit return type.
```

Comment:
Better to be added with https://github.com/JetBrains/kotlin/commit/def647c0941769e3f7e08ecf242174f74c626e80#diff-5fb4206d83fbcd3e7c827c3b12541010eaf8de08ed8dd2bd470b993cfc8760d9

Anyway, this also indicates why `Unit` return needs a special handling.